engine/apps/heartbeat/tasks.py

```diff
@@ -1,57 +1,92 @@
+        # TODO: consider migrate timeout_seconds from IntegerField to DurationField
+        .annotate(timeout=(ExpressionWrapper(F("timeout_seconds") * 10**6, output_field=DurationField())))
+    )
+    with transaction.atomic():
+        # Heartbeat is considered expired if it
+        # * is enabled,
+        # * is not already expired,
+        # * has not received a checkup for timeout period
+        expired_heartbeats = enabled_heartbeats.filter(last_heartbeat_time__lte=(timezone.now() - F("timeout"))).filter(
+            previous_alerted_state_was_life=True
+        )
+        # Schedule alert creation for each expired heartbeat after transaction commit
+        for heartbeat in expired_heartbeats:
+            transaction.on_commit(
+                lambda: create_alert.apply_async(
+                    kwargs={
+                        "title": heartbeat.alert_receive_channel.heartbeat_expired_title,
+                        "message": heartbeat.alert_receive_channel.heartbeat_expired_message,
+                        "image_url": None,
+                        "link_to_upstream_details": None,
+                        "alert_receive_channel_pk": heartbeat.alert_receive_channel.pk,
+                        "integration_unique_data": {},
+                        "raw_request_data": heartbeat.alert_receive_channel.heartbeat_expired_payload,
+                    },
+                )
+            )
+        # Update previous_alerted_state_was_life to False
+        expired_heartbeats.update(previous_alerted_state_was_life=False)
+    with transaction.atomic():
+        # Heartbeat is considered restored if it
+        # * is enabled, expired,
+        # * has received a checkup in timeout period from now,
+        # * was is alerted state (previous_alerted_state_was_life is False)
+        restored_heartbeats = (
+            enabled_heartbeats.select_for_update()
+            .filter(last_heartbeat_time__gte=(Now() - F("timeout")))
+            .filter(previous_alerted_state_was_life=False)
+        )
+        # Schedule auto-resolve alert creation for each expired heartbeat after transaction commit
+        for heartbeat in restored_heartbeats:
+            transaction.on_commit(
+                lambda: create_alert.apply_async(
+                    kwargs={
+                        "title": heartbeat.alert_receive_channel.heartbeat_restored_title,
+                        "message": heartbeat.alert_receive_channel.heartbeat_restored_message,
+                        "image_url": None,
+                        "link_to_upstream_details": None,
+                        "alert_receive_channel_pk": heartbeat.alert_receive_channel.pk,
+                        "integration_unique_data": {},
+                        "raw_request_data": heartbeat.alert_receive_channel.heartbeat_restored_payload,
+                    },
+                )
+            )
+        restored_heartbeats.update(previous_alerted_state_was_life=True)
```

Comment:
could be useful to add some test coverage here, wdyt?

Reply:
yes, of course